src/include/storage/sql_table.h

```diff
@@ -140,93 +137,229 @@ class SqlTable {
+
+  /**
+   * Translates out_buffer from desired version col_ids to tuple version col_ids, by mapping each col_id in out_buffer
+   * to its matching col_id in tuple version (2 col_ids match if they map to the same col_oid) or IGNORE_COLUMN_ID if
+   * no match exists in tuple version.
+   * @tparam RowType Type of the output (i.e. projected row, projected column)
+   * @param out_buffer a buffer initially with desired version, and translated to tuple version
+   * @param tuple_version the old schema version
+   * @param desired_version the desired schema version
+   * @param cached_orig_header original header cached
+   * @param size_map columns which have size mismatch
+   * @return whether there are columns in the desired schema version but not in the tuple version.
+   */
+  template <class RowType>
+  std::vector<std::pair<size_t, catalog::col_oid_t>> AlignHeaderToVersion(RowType *out_buffer,
+                                                                          const DataTableVersion &tuple_version,
+                                                                          const DataTableVersion &desired_version,
+                                                                          col_id_t *cached_orig_header,
+                                                                          AttrSizeMap *size_map) const;
+
+  /**
+   * Fill the missing columns in the out_buffer with default values of those columns in the desired_version
+   * @tparam RowType Type of the output (i.e. projected row, projected column)
+   * @param out_buffer a buffer that corresponds to an old schema version
+   * @param desired_version desired schema version
+   */
+  template <class RowType>
+  void FillMissingColumns(RowType *out_buffer, const std::vector<std::pair<size_t, catalog::col_oid_t>> &missingl_cols,
+                          layout_version_t tuple_version, layout_version_t layout_version) const;
+
+  /**
+   * Creates a new datatble version given the schema and version number
+   * @param schema the initial Schema of this SqlTable
+   * @param version Schema version of the created/updated table version
+   * @return If the datatable version has been successfully created
+   */
+  bool CreateTable(common::ManagedPointer<const catalog::Schema> schema, layout_version_t version);
 
   /**
    * Given a set of col_oids, return a vector of corresponding col_ids to use for ProjectionInitialization
    * @param col_oids set of col_oids, they must be in the table's ColumnMap
    * @return vector of col_ids for these col_oids
    */
-  std::vector<col_id_t> ColIdsForOids(const std::vector<catalog::col_oid_t> &col_oids) const;
+  std::vector<col_id_t> ColIdsForOids(const std::vector<catalog::col_oid_t> &col_oids,
+                                      layout_version_t layout_version = layout_version_t{0}) const;
 
   /**
    * @warning This function is expensive to call and should be used with caution and sparingly.
+   * (TODO: no longer expensive?)
```

Comment:
Is this to say it's not expensive anymore or that you're thinking of how to not make it expensive?